include/ppx/knob.h

```diff
@@ -0,0 +1,389 @@
+// Knob Classes
+// ---------------------------------------------------------------------------------------------
+
+// Knob is an abstract class which contains common features for all knobs and knob hierarchy
+class Knob
+{
+public:
+    Knob(const std::string& flagName)
+        : mFlagName(flagName), mDisplayName(flagName), mIndent(0) {}
+    virtual ~Knob() = default;
+
+    std::string GetDisplayName() const { return mDisplayName; }
+    std::string GetFlagName() const { return mFlagName; }
+    std::string GetFlagHelp() const { return mFlagDesc; }
+    size_t      GetIndent() const { return mIndent; }
+
+    // Populate flag members
+    void SetDisplayName(std::string displayName) { mDisplayName = displayName; }
+    void SetFlagDesc(std::string flagDesc) { mFlagDesc = flagDesc; }
+    void SetIndent(size_t indent) { mIndent = indent; }
+
+    virtual void Draw() = 0;
+
+    // Flag description in usage message
+    virtual std::string GetFlagHelpText() const = 0;
+
+    // Updates knob value from commandline flag
+    virtual void UpdateFromFlag(const CliOptions& opts) = 0;
+
+    // Default values are the values when the application starts. The knobs can be reset to default by a button in the UI
+    virtual void ResetToDefault() = 0;
+
+protected:
+    std::string mDisplayName;
+    std::string mFlagName;
+    std::string mFlagDesc;
+    size_t      mIndent;
+};
+
+// KnobCheckbox will be displayed as a checkbox in the UI
+// Currently supports variants of type: bool
+template <typename T, std::enable_if_t<std::is_same_v<T, bool>, bool> = true>
+class KnobCheckbox final
+    : public Knob
+{
+public:
+    KnobCheckbox(const std::string& flagName, T defaultValue)
+        : Knob(flagName)
+    {
+        SetDefault(defaultValue);
+    }
+
+    void Draw() override
+    {
+        ImGui::Checkbox(mDisplayName.c_str(), &mValue);
+    }
+
+    void ResetToDefault() override { SetValue(mDefaultValue); }
+
+    // GetFlagHelpText format:
+    // --flag_name <true|false> : flag description
+    std::string GetFlagHelpText() const override
+    {
+        std::string flagHelpText = "--" + mFlagName + " <true|false>";
+        if (!mFlagDesc.empty()) {
+            flagHelpText += " : " + mFlagDesc;
+        }
+        return flagHelpText + "\n";
+    }
+
+    // Updates mValue and mDefaultValue, expected flag format:
+    // --flag_name <true|false>
+    void UpdateFromFlag(const CliOptions& opts) override
+    {
+        SetDefault(opts.GetExtraOptionValueOrDefault(mFlagName, mValue));
+    }
+
+    T GetValue() const { return mValue; }
+
+    // Used for when mValue needs to be updated outside of UI
+    // Updates mValue
+    void SetValue(T newVal)
+    {
+        mValue = newVal;
+    }
+
+private:
+    // SetDefault will set mDefaultValue and update mValue
+    void SetDefault(T newVal)
+    {
+        mDefaultValue = newVal;
+        ResetToDefault();
```

Comment:
I think you did that for the flag update case:
 - flag changes default value, but application still showing the last default value.
 
But it seems a bit wrong to have a `SetDefault` changing the current value.
What about having the `ResetToDefault()` called in the `UpdateFromFlags`?